### PR TITLE
Get all shares iterable

### PR DESCRIPTION
**Pull request by rullzer**
Easy way to loop over all shares. For example in commands or cronjobs

### CODE REVIEW

File: lib/private/Share20/Manager.php [hunk 1]
```diff
@@ -1671,14 +1671,11 @@ public function shareProviderExists($shareType) {
...
 	}
 
-	public function getFiltered(\Closure $closure): array {
+	public function getAllShares(): iterable {
 		$providers = $this->factory->getAllProviders();
 
-		$shares = [];
 		foreach ($providers as $provider) {
-			$shares = array_merge($shares, $provider->getFiltered($closure));
+			yield from $provider->getAllShares();
```
Comment (ChristophWurst) [inline lib/private/Share20/Manager.php]: You're a magician! Didn't know this was possible. Awesome!

File: lib/public/Share/IShareProvider.php [hunk 2]
```diff
@@ -219,14 +219,10 @@ public function userDeletedFromGroup($uid, $gid);
 	public function getAccessList($nodes, $currentAccess);
 
 	/**
-	 * Get all the shares in this provider filtered with the closure
-	 * The closure requires to follow
+	 * Get all the shares in this provider returned as iterable to reduce memory
+	 * overhead
```
Comment (ChristophWurst) [inline lib/public/Share/IShareProvider.php]: Add warning here that the iterable should be iterated until the end?

File: lib/private/Share20/Manager.php (hunk 1 above)
Comment (rullzer) [reply lib/private/Share20/Manager.php]: :)

File: lib/public/Share/IShareProvider.php (hunk 2 above)
Comment (rullzer) [reply lib/public/Share/IShareProvider.php]: fine by me

**Review by skjnldsv — APPROVED** [review]
Code looks good! :D
You learn things every day in php :see_no_evil:

File: lib/private/Share20/DefaultShareProvider.php [hunk 3]
```diff
@@ -1382,4 +1382,30 @@ private function sendNote(array $recipients, IShare $share) {
 		}
 
 	}
+
+	public function getAllShares(): iterable{
```
Comment (nickvergessen) [inline lib/private/Share20/DefaultShareProvider.php]:
```suggestion
	public function getAllShares(): iterable {
```

**Comment by kesselb** [comment]
@danxuliu :+1:

**Review by danxuliu — APPROVED** [review]

File: tests/lib/Share20/ManagerTest.php [hunk 4]
```diff
@@ -3685,6 +3685,57 @@ public function testGetAccessListWithCurrentAccess() {
...
+		$extraProvider->method('getAllShares')
+			->willReturnCallback(function() use ($share3, $share4) {
+				yield $share3;
+				yield $share4;
+			});
+
+		// "yield from", used in "getAllShares()", does not reset the keys, so
+		// "use_keys" has to be disabled to collect all the values while
+		// ignoring the keys returned by the generator.
+		$result = iterator_to_array($manager->getAllShares(), $use_keys = false);
```
[the remark this answers is not in the document]
Comment (danxuliu) [reply tests/lib/Share20/ManagerTest.php]: It is not needed, but I do not like unnamed boolean parameters ;-) (I find `iterator_to_array($manager->getAllShares(), $use_keys = false)` clearer than `iterator_to_array($manager->getAllShares(), false)`).

[the remark this answers is not in the document]
Comment (danxuliu) [reply tests/lib/Share20/ManagerTest.php]: Oh, I know; I am a bit uncommon too :-P

**Pull request merged**